contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,850 @@
+ not block due to a loss of capacity or due to the scheduling of another Pod
+ between Pod termination and Pod creation. This is mitigated by blocking the 
+ update when a Pod fails to schedule. Remediation will require operator 
+ intervention. This implementation is no worse than the current behavior with 
+ respect to eviction.
+ - We will eventually implement a signal that is delivered to Pods to indicate 
+ the 
+ [reason for termination](https://github.com/kubernetes/kubernetes/issues/1462).
+ This will be a general implementation, usable for any Pod in a Kubernetes 
+ cluster. It is, therefore, out of scope to design such a mechanism here.
+ - Kubelet does not support resizing a container's resources without terminating 
+ the Pod. In place resource reallocation is out of scope for this design. 
+ Vertical scaling must be performed destructively.
+ - The primary means of configuration update will be configuration files, 
+ command line flags, environment variables, or ConfigMaps consumed as the one 
+ of the former. 
+ - In place configuration update via SIGHUP is not universally 
+ supported, and Kubelet provides no mechanism to perform this currently. Pod 
+ reconfiguration will be performed destructively.
+ - Stateful applications are likely to evolve wire protocols and storage formats
+  between versions. In most cases, when updating the application's Pod's 
+  containers, it will not be safe to roll back or forward to an arbitrary 
+  version. StatefulSet update should work well when rolling out an update, 
+  or performing a rollback, between two specific revisions of the StatefulSet.
+
+## Requirements
+This design is based on the following requirements.
+- Users must be able to update the containers of a StatefulSet's Pods.
+  - Updates to container commands, images, resources and configuration must be 
+  supported.
+- The update must progress in a sequential, deterministic order and respect the 
+  StatefulSet
+  [identity](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#pod-identity), 
+  [deployment, and scaling](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#deployment-and-scaling-guarantee) 
+  guarantees.
+- A failed update must halt.
+- Users must be able to rollback an update.
+- Users must be able to roll forward to fix a failing/failed update.
+- Users must be able to view the status of an update.
+- Users should be able to view a bounded history of the updates that have been 
+applied to the StatefulSet.
+
+## API Object
+
+The following modifications will be made to the StatefulSetStatus API object.
+
+```go
+ type StatefulSetStatus struct {
+    // ObservedGeneration and Replicas fields are ommitted for brevity.
+
+ 	// CurrentTemplateRevision, if not nil, is the revision of the PodTemplate
+ 	// that was used to create Pods with ordinals in the sequence
+    // [0,CurrentRevisionReplicas).
+ 	TemplateRevision  *int64 `json:"templateRevision,omitempty"`
```

Comment:
> Is this a usability issue with Deployments that we need to fix?

The fact that we don't output the revisions in the status, both the one we roll from and the one we roll to.